xpotato/dataset/dataset.py

```diff
@@ -44,6 +50,258 @@ def save_dataframe(df: pd.DataFrame, path: str) -> None:
+                        e2_not_found.append(t.text)
+
+            ud_graph_list.append(ud_graph.G)
+            fl_graphs.append(fl_graph.G)
+
+            # Build a report of missing or correctly mapped tokens for the FL graph
+            reports.append(f"Entity 1: Found {tokens_e1_found} / {tokens_e1} token nodes\n" +
+                           f"{entity1_dict}\n" +
+                           f"Not found: {e1_not_found}\n\n" +
+                           f"Entity 2: Found {tokens_e2_found} / {tokens_e2} token nodes\n" +
+                           f"{entity2_dict}\n" +
+                           f"Not found: {e2_not_found}\n\n")
+            e1_found.append(0.0 if tokens_e1 == 0 else tokens_e1_found / tokens_e1)
+            e2_found.append(0.0 if tokens_e2 == 0 else tokens_e2_found / tokens_e2)
+
+        df_parsed = pd.DataFrame(
+            {
+                "SID": sentence_ids,
+                "text": sentences,
+                "label_id": label_ids,
+                "ud": ud_graph_list,
+                "fl": fl_graphs,
+                "report_fl": reports,
+                "e1_found_fl": e1_found,
+                "e2_found_fl": e2_found
+            }
+        )
+
+        return df_parsed
+
+    @staticmethod
+    def generate_dataframe_amr(df, sentence_colname="", label_id_colname="", sentence_id_colname=""):
+        """
+        Submit a dataframe and compute a response dataframe with entity-marked graphs
+        Expects columns b1,e1 b2,e2 to provide begin/end character positions of the respective entities 1 and 2.
+        :param df: The dataframe containing the input sentence and all necessary conversion information
+        :param sentence_colname: which column is the sentence in
+        :param label_id_colname: which column is the label in
+        :param sentence_id_colname: which column is the unique sentence id in
+        :return: a new dataframe with the above information, converted and entity-tagged AMR graphs as well as a
+                 report for how many and which entity tokens could be mapped to their respective AMR nodes
+        """
+        def is_allcaps(text):
+            return not re.search(r'[a-z]', text)
+
+        # For each graph keep track of which entity tokens are linked to nodes, and which couldn't be found
+        rows = df.iterrows()
+        sentences = [row[1][sentence_colname] for row in rows]
+        rows = df.iterrows()
+        label_ids = [row[1][label_id_colname] for row in rows]
+        rows = df.iterrows()
+        sentence_ids = [row[1][sentence_id_colname] for row in rows]
+
+        amr_graph_list = []
+        reports = []
+        e1_found = []
+        e2_found = []
+
+        # For my conversions I have been using the amrlib 0.8.0 xfm bart large model
+        amr_stog = amrlib.load_stog_model()
+        # Load the same spacy model that TUW NLP uses to build the graphs to get a comparable mapping
+        spacy_nlp = spacy.load('en_core_web_sm')
```

Comment:
It feels like we are loading/doing the parsing redundantly. I think if this information is not available from the AMR graph the TUW library provides, we should modify that code instead and just call it here.

Reply:
Yes, then we would need to adapt the AMRGraph class itself. Currently, it does not provide a link from token to character position in the source text, and that is generally how entities are marked. That's why the spacy method is called again, so we can reproduce the source text to token mapping that is usually done within the AMR Graph conversion process.